### PR TITLE
refactor checkbox with dependant-group trait

### PR DESCRIPTION
**Pull request by nhaagen**
The code of Group is used in Checkbox, e.g., but imo a checkbox is not a group. Thus, classes should rather share a trait. This also straightens rendering.

Please consider as kind suggestion ;)

### CODE REVIEW

File: src/UI/Component/Input/Field/Checkbox.php [hunk 1]
```diff
@@ -11,23 +11,6 @@
 /**
  * This describes checkbox inputs.
  */
-interface Checkbox extends Group {
+interface Checkbox extends Input {
```
Comment (klees) [inline src/UI/Component/Input/Field/Checkbox.php]: Shouldn't this be `DependantGroupProviding` as well?

File: src/UI/Component/Input/Field/DependantGroupProviding.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	 *
+	 * @return Input
+	 */
+	public function withDependantGroup(DependantGroup $dependant_group);
+
+
+	/**
+	 * Returns the attached DependantGroup or null if none is attached.
+	 *
+	 * @return $dependantGroup|null
```
Comment (klees) [inline src/UI/Component/Input/Field/DependantGroupProviding.php]: `$dependantGroup` is not a type.

File: src/UI/Component/Input/Field/DependantGroupProviding.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ */
+interface DependantGroupProviding {
+
+	/**
+	 * Creates an input like this but with a dependant group attached which appears if the
+	 * control is clicked.
+	 *
+	 * @param DependantGroup $dependant_group group to be attached to the checkbox
+	 *
+	 * @return Input
```
Comment (klees) [inline src/UI/Component/Input/Field/DependantGroupProviding.php:18]: Move these to PHP7 return type hints.

File: src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace ILIAS\UI\Implementation\Component\Input\Field;
+
+use ILIAS\UI\Component as C;
+
+/**
+ * This is a trait for inputs providing dependant groups, such as checkboxes, e.g.
+ */
+trait DependantGroupProvider {
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php]: Lets name this `DependantGroupHelper`, most traits seemed to be named `Helper` atm.

File: src/UI/Implementation/Component/Input/Field/Checkbox.php [hunk 5]
```diff
@@ -14,39 +14,23 @@
...
-	public function __construct(
-		DataFactory $data_factory,
-		ValidationFactory $validation_factory,
-		TransformationFactory $transformation_factory,
-		$label,
-		$byline
-	) {
-		parent::__construct($data_factory, $validation_factory, $transformation_factory, [], $label, $byline);
+	protected function getConstraintForRequirement() {
+		throw new \LogicException("NYI: What could 'required' mean here?");
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Checkbox.php]: Don't do this. This will crash. Just return `null`.

File: src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	public function appendOnChange(C\Signal $signal) {
+		return $this->appendTriggeredSignal($signal, 'change');
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function withOnLoad(C\Signal $signal) {
+		return $this->addTriggeredSignal($signal, 'load');
+	}
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php]: How are these related to the `DependantGroup`?

**Review by klees — CHANGES_REQUESTED** [review]
LGTM in general, some room for improvement. Thx!

File: src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php (hunk 6 above)
Comment (nhaagen) [reply src/UI/Implementation/Component/Input/Field/DependantGroupProvider.php]: DependantGroup needs a signal from owning input, e.g. on "hide/show on init" or "checkbox clicked".
The providing input must relay/translate the signal the group.

File: src/UI/Component/Input/Field/Checkbox.php (hunk 1 above)
Comment (nhaagen) [reply src/UI/Component/Input/Field/Checkbox.php]: I like the way the class-declaration looks:
  class Checkbox extends Input implements C\Input\Field\Checkbox, C\Changeable, C\Onloadable, C\Input\Field\DependantGroupProviding 

with 
  interface Checkbox extends DependantGroupProviding 
we would have to stick to
  class Checkbox extends Input implements C\Input\Field\Checkbox, C\Changeable, C\Onloadable 
...which is fine, but - imo - not as nice.

Comment (klees) [reply src/UI/Component/Input/Field/Checkbox.php]: But this would also mean, that this is something that only is available internally, users of the UI-Framework won't see this. This of course is also true for `Changeable` and `Onloadable`. Is this what we want?

Comment (Amstutz) [reply src/UI/Component/Input/Field/Checkbox.php]: I think on public would be nicer. However, no strong opinion here.

**Review by Amstutz — APPROVED** [review]
Thanks a lot for this, a good step forward for the inputs. Did you run the tests? Do they still pass? If so, I am fine with the changes, if not ...

**Comment by nhaagen** [comment]
all tests pass...
